### PR TITLE
Fusion

### PR DESCRIPTION
**Pull request by FlaminGuy**
deprecations were mentioned in #513.

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/runtime/Benchmark/FlixListInScala.scala [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+// Implementation of Flix List in Scala
+object FlixListInScala {
+  // This trait represents `List` enum in Flix.
+  sealed trait FlixList[+T]
+
+  // This object represents `Nil` case in `List`
+  case object FlixNil extends FlixList[Nothing]
+
+  // This case class represents `Cons` case in `List`
+  case class FlixCons[+T](field: (T, FlixList[T])) extends FlixList[T]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/Benchmark/FlixListInScala.scala]: Ok, this is definitely apples-to-apples. But for it to be idiomatic scala, we have to change it to:

```
case class FlixCons(x: T, rs: FlixList[T])
```

Otherwise people will argue that we are giving ourselves an unfair advantage. 

If you are wondering what we want to compare ourselves to, it should be this: 

"Idiomatic, but purely functional, Scala code".

**Comment by magnus-madsen** [comment]
I added an important comment.

We might not merge this into the master branch, since it has nothing to do with the Flix compiler/language itself. But having the PR is still useful for code reviews etc.

**Comment by FlaminGuy** [comment]
I think this should go into master because it's useful and convenient to have it there, but I also think we should put it in another directory and not include it in binaries or things like that.

**Comment by FlaminGuy** [comment]
This used to be the benchmarks, but I took them out since they shouldn't be in my master branch. 
From the other review:
We create java classes representing a class which is a fusion of both tuples and enum cases.
Fusion classes include all methods in both tuple and enum case. When we generate an enum which has a tuple as it's field, we will generate a fusion object instead of generating one tuple object and one enum object as an optimization.
This phase is an optimization since we will generate less objects which helps garbage collector and we will spend less time in generating these objects.
This pull request should get merged in when we fix the issue for toString() and equals() methods on generated objects which shouldn't be called for this phase to work.
Comments on the other review are all fixed.

**Comment by FlaminGuy** [comment]
There is still a bug here due to a dependency on hashCodes for generated objects.

File: main/src/ca/uwaterloo/flix/Main.scala [hunk 2]
```diff
@@ -19,7 +19,8 @@ package ca.uwaterloo.flix
 import java.io.File
 
 import ca.uwaterloo.flix.api.{Flix, MatchException, RuleException, SwitchException, UserException}
-import ca.uwaterloo.flix.runtime.{Benchmarker, Shell, Tester, Value}
+import ca.uwaterloo.flix.runtime.Benchmark.Benchmarker
+import ca.uwaterloo.flix.runtime._
 import ca.uwaterloo.flix.util._
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/Main.scala]: Unnecessary changes. (Main is unchanged).

File: main/src/ca/uwaterloo/flix/language/ast/ExecutableAst.scala [hunk 3]
```diff
@@ -42,11 +42,12 @@ object ExecutableAst {
...
                   dependenciesOf: Map[Symbol.TableSym, Set[(Constraint, ExecutableAst.Predicate.Body.Positive)]]) extends ExecutableAst
 
-  case class ByteCodes(enumInterfaceByteCodes: Map[EnumSym, (QualName, Array[Byte])],
+  case class ByteCodes(enumInterfaceByteCodes:  Map[EnumSym, (EnumTypeInterfaceName, Array[Byte])],
                        enumClassByteCodes: Map[EnumSym, (Map[(String, Type), Array[Byte]], Map[String, Array[Byte]])],
                        functionalInterfaceByteCodes: Map[Type, (FlixClassName, Array[Byte])],
                        classByteCodes: Map[FlixClassName, Array[Byte]],
-                       tupleByteCode: Map[TupleClassName, Array[Byte]])
+                       tupleByteCode: Map[List[WrappedType], (Array[Byte], Array[Byte])],
+                       ETFusionByteCode: Map[EnumSym, Map[String, Map[List[WrappedType], Array[Byte]]]])
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ExecutableAst.scala]: At some point we need to refactor all of this. What is the most general map that can hold all of these?

A Flix type is mapped to a Java type and a Java type has a name (interface or class), so could we do with:

```
Map[JavaName, Array[Byte]]
```
?

File: main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala [hunk 4]
```diff
@@ -273,9 +274,10 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
                                 functions: List[ExecutableAst.Def],
                                 declarations: Map[Symbol.DefnSym, Type],
                                 interfaces: Map[Type, FlixClassName],
-                                enums: Map[(Type, String), (QualName, ExecutableAst.Case)],
+                                enums: Map[(Type, String), (ExecutableAst.Case, Type)],
                                 visitor: MethodVisitor,
-                                entryPoint: Label)(expr: Expression): Unit = expr match {
+                                entryPoint: Label,
+                                options: Options)(expr: Expression): Unit = expr match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: This signature is getting really long. I think it can be collapsed like so:

- The functions are available in the AST root?
- The declarations are also available in the AST root?
- The interfaces and enums looks like something that can be grouped into one algebraic data type.
- The visitor is necessary.
- The entryPoint is necessary.
- The options is necessary (but might as well be accessed from the Flix object).

In other Flix phases we are starting to use this pattern (basically just nested functions):

```
def whaterver(root: Root)(implicit flix: Flix): .. {

  def doStuffWithExp(exp: Exp, ...) {
   // notice that in here both the AST root and the Flix object is available.
  }
}
```

This structure nicely "emulates" a form of global variables in a phase, but in a purely functional way using lexical scope.

File: main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala [hunk 5]
```diff
@@ -100,49 +100,107 @@ object CodegenHelper {
...
   case class TupleClassName(fields: List[WrappedType]) extends QualName {
     val ref: List[String] = fixedTuplePrefix ::: fields.map {
       case WrappedPrimitive(tpe) => typeSpecifier(tpe)
       case WrappedNonPrimitives(_) => "object"
-    } ::: List("Tuple")
+    } ::: List("clazz", "Tuple")
   }
 
-  /**
+  def exceptionThrowerMethod(visitor: ClassWriter,
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala]: This is a method with a reasonable number of arguments that are understandable.

File: main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala [hunk 6]
```diff
@@ -191,6 +249,40 @@ object CodegenHelper {
...
 
   /**
+    * At this method, we group `tuples` that have the same field representation so we only generate one class for them.
+    * If a field is a primitive, then it can be represented by it's primitive but if the field is not a primitive then it
+    * has to be represented using an object.
+    * For example, `(List[Int32], Bool)` and `(Result[Int32,Int32], Bool)` have the same representation since the first field
+    * of both of them is an object and the second field is a `Bool`. So we only create one tuple class for both of these
+    * tuples.
+    */
+  def groupTuplesByFieldTypes(tuples: List[Type]): List[List[List[Type]]] = tuples.map{
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala]: So the signature here is hard to understand, triple nested lists?

What I think need is a representation of Java types then you can map one Flix type into its Java type, and if you have a set of Flix types you can map them (a possible smaller) set of Java types (which is really what is happening here).

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala]: The comment was helpful though.

File: main/src/ca/uwaterloo/flix/runtime/Benchmark/Benchmarker.scala [hunk 7]
```diff
@@ -122,4 +124,4 @@ object Benchmarker {
     Thread.sleep(500)
   }
 
-}
+}
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/Benchmark/Benchmarker.scala]: Benchmarker should not be changed.

File: main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala [hunk 8]
```diff
@@ -57,11 +57,14 @@ object LoadBytecode extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
     * 5. Load Enum classes
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala]: I think this entire class can be replaced by two for loops. One for loop over a very general structure to load every generated interface and class. And one for loop over each definition symbol to load its associated method (via a helper function that maps a definition symbol to a java name),

**Comment by magnus-madsen** [comment]
I added some comments, but they are sort of higher-level and more to be discussed.

File: main/src/ca/uwaterloo/flix/language/ast/ExecutableAst.scala (hunk 3 above)
Comment (FlaminGuy) [reply main/src/ca/uwaterloo/flix/language/ast/ExecutableAst.scala]: Sure, can you create a ticket for it?

File: main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala (hunk 4 above)
Comment (FlaminGuy) [reply main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala]: Yes, functions and declarations are available in AST root, even enums and interfaces are from root. The problem is, you need to traverse the whole tree to extract these informations. We can talk about these later.

File: main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala (hunk 5 above)
Comment (FlaminGuy) [reply main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala]: That sound sarcastic in many ways :D Anyway, I added a comment there just to keep the format.

File: main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala (hunk 8 above)
Comment (FlaminGuy) [reply main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala]: We should talk about this and add some tickets for it!

**Comment by FlaminGuy** [comment]
I fixed most of the problems. 
@magnus-madsen it is still pending that change to remove dependency on `hashCode` of generated objects.
